### PR TITLE
Inference: Flashinfer bounded decode inference path for BF16

### PR DESCRIPTION
**Pull request by mathemakitten**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Direct extension of #6864 for bf16. Preserves refit buffers. 

Nemotron 3B hybrid MoE, EP4, concurrency 128:

* torch grouped GEMM, BF16: 0.0596 ms/token overall, 0.09348 decode ms/token
* FlashInfer BF16, capacity 512: 0.0506 ms/token overall, 0.06701 decode ms/token

~15% faster e2e and +28% specifically in decode.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py [hunk 1]
```diff
@@ -759,13 +759,13 @@ def _dynamic_step_forward_logits(self, input_ids: Tensor, position_ids: Tensor):
...
-        token_capacity = self.model_config.inference_flashinfer_mxfp8_token_capacity
+        token_capacity = self.model_config.inference_flashinfer_token_capacity
         decode_token_upper_bound = (
             context.max_requests
             * (self.num_speculative_tokens + 1)
             * self.model_config.expert_model_parallel_size
         )
-        InferenceMode.set_bounded_mxfp8_rows(
+        InferenceMode.set_bounded_flashinfer_rows(
             context.is_decode_only()
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/text_generation_controllers/text_generation_controller.py]: This bound only makes sense if all ranks in the EP group are in decode mode, and not just this rank.

**Review by sidsingh-nvidia — CHANGES_REQUESTED** [review]

**Review by yaox12 — APPROVED** [review]
